src/main/java/nextstep/subway/line/domain/Sections.java

```diff
@@ -19,49 +20,95 @@
+        Optional<Section> nextSection = findNextSection(startSection.getDownStation());
+        while (nextSection.isPresent()) {
+            stations.add(nextSection.get().getDownStation());
+            nextSection = findNextSection(nextSection.get().getDownStation());
         }
         return stations;
     }
 
-    private Section getStartSection() {
-        Section ret = null;
-        for (Section section : this.sections) {
-            ret = isNotContainDownStations(section.getUpStation());
-        }
-        return ret;
+    private void addUptoUpStation(Section section) {
+        sections.stream()
+                .filter(oldSection -> section.getUpStation() == oldSection.getUpStation())
+                .findFirst()
+                .ifPresent(oldSection -> {
+                    sections.add(new Section(oldSection.getLine(), section.getDownStation(), oldSection.getDownStation(), getUpStationDistance(section, oldSection)));
+                    sections.remove(oldSection);
+                });
     }
 
-    private Section isNotContainDownStations(Station upStation) {
-        return this.sections
-                .stream()
-                .filter(section -> isNotEqualsDownStation(section, upStation))
+    private void addDownToDownStation(Section section) {
+        sections.stream()
+                .filter(oldSection -> section.getUpStation() == oldSection.getDownStation())
                 .findFirst()
-                .orElseThrow(() -> new IllegalArgumentException("순환 구간은 없습니다."));
+                .ifPresent(oldSection -> {
+                    sections.add(new Section(oldSection.getLine(), oldSection.getUpStation(), section.getUpStation(), section.getDistance().get()));
+                    sections.remove(oldSection);
+                });
     }
 
-    private boolean isNotEqualsDownStation(Section section, Station upStation) {
-        return !section.getDownStation().getId().equals(upStation.getId());
+    private long getUpStationDistance(Section section, Section oldSection) {
+        long nextDistance = oldSection.getDistance().get() - section.getDistance().get();
+        if (nextDistance < 1) {
+            throw new IllegalArgumentException("기존 역 사이 길이보다 크거나 같으면 등록을 할 수 없음");
+        }
+        return nextDistance;
     }
 
-    private Section getNextSection(Station downStation) {
-        return this.sections
-                .stream()
-                .filter(section -> isNextStation(section, downStation))
+    private Section findStartSection() {
+        return sections.stream()
+                .filter(s -> !isContainsDownStation(s.getUpStation()))
                 .findFirst()
-                .orElse(null);
+                .orElseThrow(RuntimeException::new);
+    }
+
+    private boolean isContainsDownStation(Station station) {
+        boolean flag = false;
+        for (Section section : sections) {
+            flag = section.getDownStation() == station;
+        }
+        return flag;
```

Comment:
저번에 아래와 같은 코멘트 남겨드렸습니다.

```
for loop을 여러번 돌면서 ret 값을 바꿔주게 되면 결국 순서에 영향을 받게 될 것 같은데, 제가 잘못 이해한거라면 코멘트 남겨주시고 해당 부분 확인 부탁드리겠습니다.
```

이번에도 동일하게 flag 값이 계속 바뀌면서 section의 순서에 영향을 받을 것 같습니다.
파라미터로 들어온 upStation이 downStation에 존재하는 경우가 있더라도.. for loop는 끝까지 도니 마지막 for loop에서 downStation과 파라미터로 들어온 역이 다른 경우 원하는 값과 다른 값이 반환되지 않나요?

제가 놓친 부분이 있을 수 있는데.. github 코드 상으로 봤을 때 조금 이상한 것 같습니다.
제가 생각하기에 의도하신 내용을 구현하기에 아래 코드가 조금 더 적합하지 않나? 하는 생각이 듭니다.

```suggestion
 return this.sections.stream()
                .anyMatch((section) -> section.getDownStation() == station);
```

제가 혹시 코드를 잘못 이해하고 코멘트를 남긴거라면 메세지 남겨주세요